### PR TITLE
[SPARK-38140][SQL] Desc column stats (min, max) for timestamp type is not consistent with the values due to time zone difference

### PR DESCRIPTION
**Pull request by wzhfy**
### What changes were proposed in this pull request?

Currently timestamp column's stats (min/max) are stored using UTC time zone in metastore, and when desc its min/max column stats, they are also shown in UTC.
As a result, for users not in UTC, the column stats (shown to users) are not consistent with the actual value, which causes confusion.
Note that it does not affect correctness. But we'd better to remove confusion for users.

### Why are the changes needed?

To make column stats and column value consistent when shown to users.

### Does this PR introduce _any_ user-facing change?

As an example:
```
spark-sql> create table tab_ts_master (ts timestamp) using parquet;
spark-sql> insert into tab_ts_master values make_timestamp(2022, 1, 1, 0, 0, 1.123456), make_timestamp(2022, 1, 3, 0, 0, 2.987654);
spark-sql> select * from tab_ts_master;
2022-01-01 00:00:01.123456
2022-01-03 00:00:02.987654
spark-sql> set spark.sql.session.timeZone;
spark.sql.session.timeZone	Asia/Shanghai
spark-sql> analyze table tab_ts_master compute statistics for all columns;
```
Before this change:
```
spark-sql> desc formatted tab_ts_master ts;
col_name	ts
data_type	timestamp
comment	NULL
min	2021-12-31 16:00:01.123456
max	2022-01-02 16:00:02.987654
num_nulls	0
distinct_count	2
avg_col_len	8
max_col_len	8
histogram	NULL
```
The min/max column stats are inconsistent with what the user sees in the column values.

After this change:
```
spark-sql> desc formatted tab_ts ts;
col_name	ts
data_type	timestamp
comment	NULL
min	2022-01-01 00:00:01.123456
max	2022-01-03 00:00:02.987654
num_nulls	0
distinct_count	2
avg_col_len	8
max_col_len	8
histogram	NULL
```
### How was this patch tested?

Added new unit tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 1]
```diff
@@ -787,7 +789,22 @@ case class DescribeColumnCommand(
       } yield histogramDescription(hist)
       buffer ++= histDesc.getOrElse(Seq(Row("histogram", "NULL")))
     }
-    buffer.toSeq
+    buffer
```
Comment (sarutak) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:792]: Please keep it `buffer.toSeq` for Scala 2.13.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 2]
```diff
@@ -787,7 +789,22 @@ case class DescribeColumnCommand(
...
+  }
+
+  private def toZoneAwareExternalString(s: String, name: String, dataType: DataType): String = {
```
Comment (martin-g) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:795]: please use better variable name for `s`

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 3]
```diff
@@ -470,7 +470,98 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
...
+    }
+  }
+
+  test("describe column stats (min, max) for timestamp column: desc should show different " +
+    "results if writing in UTC and desc in other time zones") {
+
+    val table = "insert_desc_diff_time_zones"
+    val tsCol = "timestamp_typed_col"
+
+    val original = TimeZone.getDefault
```
Comment (sarutak) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala:527]: Can we use `withDefaultTimeZone` here?

Comment (wzhfy) [reply sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: Thanks! Didn't know we have this util method. Updated.

**Comment by cloud-fan** [comment]
I agree that using local timezone to display the min/max timestamp values is more user-friendly, but I'd suggest to put timezone at the end, to avoid confusion, e.g. `2022-01-01 00:00:01.123456 +8:00`

**Comment by yaooqinn** [comment]
will this break the consistency of the stats between historical and incremental data?

**Comment by wzhfy** [comment]
> will this break the consistency of the stats between historical and incremental data?

What do you mean by consistency? The internal value (long type) is not changed, it only changes the exposed result (e.g. by desc command).

**Comment by wzhfy** [comment]
> I agree that using local timezone to display the min/max timestamp values is more user-friendly, but I'd suggest to put timezone at the end, to avoid confusion, e.g. `2022-01-01 00:00:01.123456 +8:00`

Good point. Thanks!

**Comment by wzhfy** [comment]
kindly ping @sarutak @martin-g @cloud-fan @yaooqinn

**Review by cloud-fan — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**